controllers/user_controller.go

```diff
@@ -0,0 +1,428 @@
+	OverrideExistingSecret bool
+}
+
+const (
+	userFinalizer = "s3.onyxia.sh/finalizer"
+)
+
+//+kubebuilder:rbac:groups=s3.onyxia.sh,resources=S3User,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=s3.onyxia.sh,resources=S3User/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=s3.onyxia.sh,resources=S3User/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *S3UserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// Checking for userResource existence
+	userResource := &s3v1alpha1.S3User{}
+	err := r.Get(ctx, req.NamespacedName, userResource)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info(fmt.Sprintf("S3User CRD %s has been removed. NOOP", req.Name))
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Check if the userResource instance is marked to be deleted, which is
+	// indicated by the deletion timestamp being set. The object will be deleted.
+
+	if userResource.GetDeletionTimestamp() != nil {
+		logger.Info("userResource have been marked for deletion")
+		return handleS3UserDeletion(ctx, userResource, r, logger, err)
+	}
+
+	// Check user existence on the S3 server
+	found, err := r.S3Client.UserExist(userResource.Spec.Name)
+	if err != nil {
+		logger.Error(err, "an error occurred while checking the existence of a user", "user", userResource.Spec.Name)
+		return r.setS3UserStatusConditionAndUpdate(ctx, userResource, "OperatorFailed", metav1.ConditionFalse, "S3UserExistenceCheckFailed",
+			fmt.Sprintf("Obtaining user[%s] info from S3 instance has failed", userResource.Spec.Name), err)
+	}
+
+	// Add finalizer for this CR
+	if !controllerutil.ContainsFinalizer(userResource, userFinalizer) {
+		controllerutil.AddFinalizer(userResource, userFinalizer)
+		err = r.Update(ctx, userResource)
+		if err != nil {
+			logger.Error(err, "an error occurred when adding finalizer from user", "user", userResource.Spec.Name)
+			// return ctrl.Result{}, err
+			return r.setS3UserStatusConditionAndUpdate(ctx, userResource, "OperatorFailed", metav1.ConditionFalse, "S3UserFinalizerAddFailed",
+				fmt.Sprintf("An error occurred when attempting to add the finalizer from user [%s]", userResource.Spec.Name), err)
+		}
+	}
+
+	// If the user does not exist, it is created based on the CR
+	if !found {
+		// S3User creation
+		// The user creation happened without any error
+		return handleS3UserCreation(ctx, userResource, r)
+	}
+
+	// if the user exist, update from the CR
+	if found {
+		// TODO Check if secret still exist and is valid
+		// TODO Delete S3User and create new one
+		//TODO Manage error
+		//TODO Manage error
```

Comment:
That's a lot of TODO, were they actually implemented ? If so, can you remove the TODO comments from the code ?